data_model/src/predicate.rs

```diff
@@ -3,38 +3,66 @@
 #[cfg(not(feature = "std"))]
 use alloc::vec;
 
+use core::fmt::Display;
+
 use super::*;
 use crate::{IdBox, Name, Value};
 
 /// Predicate combinator enum.
 #[derive(Debug, Clone, Serialize, Deserialize, Encode, Decode, IntoSchema)]
-pub enum PredicateBox {
+pub enum GenericPredicateBox<P> {
     /// Logically `&&` the results of applying the two predicates.
-    And(Vec<PredicateBox>),
+    And(Vec<Self>),
     /// Logically `||` the results of applying the two predicates.
-    Or(Vec<PredicateBox>),
+    Or(Vec<Self>),
     /// Negate the result of applying the predicate.
-    Not(Box<PredicateBox>),
+    Not(Box<Self>),
     /// The raw predicate that must be applied.
-    Raw(value::Predicate),
+    Raw(P),
 }
 
-impl PredicateBox {
+impl<T> Display for GenericPredicateBox<T>
```

Comment:
Please omit this for now.

Reply:
Note that the standard library doesn't support `core::fmt::Display` for `alloc::vec::Vec`. This is intentional.

Reply:
That is unfortunately needed because of [this](https://github.com/hyperledger/iroha/blob/b2769df18bce2295cf39192ee2f8b3949946a4e6/core/src/smartcontracts/isi/permissions/mod.rs#L35).
I would argue that requiring `Display` isn't the optimal choice there either, but that warrants its own discussion and is out of scope for this PR.

Reply:
> I would argue that requiring Display isn't the optimal choice there either, but that warrants its own discussion and is out of scope for this PR.

Agree. In fact supertraits seem to be used as shorthands for trait combinations a whole lot in Iroha. We should address this in the style guide.